### PR TITLE
Changed chrome extension to self-host.

### PR DESCRIPTION
**Pull request by umop**
Sets up chrome extension to self host as filesystem (not yet providing executable command).

@rpaquay @ussuri

### CODE REVIEW

File: chrome_agent/manifest.json [hunk 1]
```diff
@@ -11,17 +11,16 @@
 
   "background": {
     "persistent": true,
-    "scripts": [
-      "scripts/agent_background.js"
-    ]
+    "page": "background.html"
```
Comment (ussuri) [inline chrome_agent/manifest.json]: `"page"` is unofficially deprecated, and having background.html looks weird for an extension anyway (even for an app). Why not just list the scripts imported in background.html here?

Comment (umop) [inline chrome_agent/manifest.json]: sgtm.  Will change

File: lib/axiom/fs/stream/post_message_streams.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  this.handleReceive_(message);
+};
+
+/**
+ * Return the state of the connection.
+ *
+ * @protected
+ */
+PostMessageStreams.prototype.getState_ = function() {
+  if (this.port == null) {
```
Comment (ussuri) [inline lib/axiom/fs/stream/post_message_streams.js]: `this.port === null`. With `==`, `undefined == null` evaluates to `true` (just one example).

Comment (umop) [inline lib/axiom/fs/stream/post_message_streams.js]: Good catch.  Thanks.  Done.

**Comment by umop** [comment]
Two additional notes:

@rpaquay I included the refactoring of WebSocketStreams in here, but I'll revert if you prefer.  It doesn't need to subclass GenericStreams.

@ussuri I copied the full chrome_api.js to externs with some additional missing references.  It did look like the file had been intentionally paired down, though, so I can prune out unused externs if there is a reason to keep the file small (I figured there wasn't since it is not used during compilation, but just during checking).  LMK, thanks!

File: chrome_agent/js/chrome_agent.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import Transport from 'axiom/fs/stream/transport';
+import Channel from 'axiom/fs/stream/channel';
+import ExtensionStreams from 'axiom/fs/stream/extension_streams';
+
+/**
+ * Host filesystem with chrome handling executable for use with axiom.
+ *
+ * @constructor
+ */
+export var ChromeAgent = function() {
```
Comment (rpaquay) [inline chrome_agent/js/chrome_agent.js]: I would leave the `ChromeAgent` constructor body empty, and move the current body to a new method (called `connect` maybe?) that return a promise.

The issue with putting this code inside the constructor is that 1) the caller doesn't know when the agent is ready and 2) the caller does not get notified of promise failures.

File: chrome_agent/js/extension/background.js [hunk 4]
```diff
@@ -12,6 +12,12 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+__axiomExport__(this);
+var chromeAgent = new chrome_agent.ChromeAgent();
```
Comment (rpaquay) [inline chrome_agent/js/extension/background.js]: See comment above.
``` javascript
var chromeAgent = new chrome_agent.ChromeAgent();
chromeAgent.connect().then(
  function() { console.log ('Successfully connected to ChromeAgent'); }
  function(error) { console.error ('Error connecting to ChromeAgent', error); }
)
```

File: gruntfile.js [hunk 5]
```diff
@@ -237,7 +237,27 @@ module.exports = function(grunt) {
           src: ['**'],
           dest: 'tmp/samples'
         }]
-      }
+      },
+      chrome_agent: {
+        files: [{
+          expand: true,
+          cwd: 'chrome_agent',
+          src: ['**'],
```
Comment (rpaquay) [inline gruntfile.js]: Are you sure there is no extra files in there we don't want to copy (i don't know, i am asking :))

Comment (rpaquay) [inline gruntfile.js]: Alternative is to explicitly list all the file extensions...

File: lib/axiom/fs/stream/extension_streams.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ * Open an ExtensionStream at given appId
+ *
+ * @param {string} appId The appId where receiving ExtensionStream lives.
+ * @return {Promise}
+ */
+ExtensionStreams.prototype.openExtension = function(appId) {
+  this.appId = appId;
+
+  var port = chrome.runtime.connect(appId);
+  return PostMessageStreams.prototype.open.call(this, {port: port, alreadyConnected: false});
```
Comment (rpaquay) [inline lib/axiom/fs/stream/extension_streams.js]: Why not `this.open`. Also, should `PostMessageStreams.open` be `protected`?

File: lib/axiom/fs/stream/extension_streams.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * Listen (as an extension) for a connection.
+ *
+ * @param {string} appId
+ * @return {Promise}
+ */
+ExtensionStreams.prototype.listenAsExtension = function() {
+  return new Promise(function(resolve, reject) {
+    chrome.runtime.onConnectExternal.addListener(function(port) {
+      PostMessageStreams.prototype.open.call(this, {port: port, alreadyConnected: true});
```
Comment (rpaquay) [inline lib/axiom/fs/stream/extension_streams.js]: Same questions as above.

File: lib/axiom/fs/stream/generic_streams.js [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  this.onClose.fire();
+}
+
+/**
+ * Return the state of the connection.
+ *
+ * @return {GenericStreams.ConnectionState}
+ * @protected
+ */
+GenericStreams.prototype.getState_ = function() {
```
Comment (rpaquay) [inline lib/axiom/fs/stream/generic_streams.js]: Add `abstract();` in the body? (see `axiom/fs/base` for examples)

File: lib/axiom/fs/stream/generic_streams.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  }
+};
+
+/**
+ * @protected
+ * @param {*} data
+ * @return {void}
+ */
+GenericStreams.prototype.handleSend_ = function(item) {
+};
```
Comment (rpaquay) [inline lib/axiom/fs/stream/generic_streams.js]: Add abstract(); in the body? (see axiom/fs/base for examples)

File: lib/axiom/fs/stream/post_message_streams.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+ * @constructor
+ * @extends {GenericStreams}
+ */
+export var PostMessageStreams = function() {
+  GenericStreams.call(this);
+
+  /** @type {chrome.runtime.Port} */
+  this.port = null;
+  /** @private @type {?Function} */
+  this.onOpenResolve_ = null;
```
Comment (rpaquay) [inline lib/axiom/fs/stream/post_message_streams.js]: Why not use a single `Completer` to replace both onOpenResolve and onOpenReject?

File: lib/axiom/fs/stream/generic_streams.js [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+  this.callback = callback;
+};
+
+/**
+ * Handle a ReadableStream and a WritableStream over a generic connection (abstract).
+ *
+ * @constructor
+ * @implements {StreamsSource}
+ */
+export var GenericStreams = function() {
```
Comment (rpaquay) [inline lib/axiom/fs/stream/generic_streams.js]: nit (can be done later): I would rename this to `BufferedStreams` because
* `GenericStreams` is not very specific and
* The implementation is buffering both read and writes when the stream is paused and that is only one specific implementation choice of many possibilities (in the future, we should be able to not buffer locally, but consume only as data comes from the other end).

File: lib/axiom/fs/stream/post_message_streams.js [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+  this.onDisconnect_ = this.onDisconnect_.bind(this);
+};
+
+PostMessageStreams.prototype = Object.create(GenericStreams.prototype);
+
+export default PostMessageStreams;
+
+/**
+ * @return {Promise}
+ * @param {*} argObject
```
Comment (rpaquay) [inline lib/axiom/fs/stream/post_message_streams.js]: Why not giving a more precise declaration, such as
```
@param { {port: chrome.runtime.Port, alreadyConnected: boolean} } argObject
```

File: lib/wash/exe/mount.stream.js [hunk 13]
```diff
@@ -177,23 +179,29 @@ var CreateStreamsResult;
...
-  // TODO(rpaquay): Add support for more stream types.
-  return Promise.reject(
-      new AxiomError.Invalid('stream-type', streamType));
+  var streams;
+
+  return Promise.resolve().then(function() {
+    switch (streamType) {
+      case 'websocket':
+        streams = new WebSocketStreams();
+        return streams.open(streamSrc).then(function() {return "ws"});
```
Comment (rpaquay) [inline lib/wash/exe/mount.stream.js]: ; after return

File: lib/wash/exe/mount.stream.js [hunk 14]
```diff
@@ -177,23 +179,29 @@ var CreateStreamsResult;
...
+      case 'extension':
+        streams = new ExtensionStreams();
+        return streams.openExtension(streamSrc).then(function() {return "ext"});
```
Comment (rpaquay) [inline lib/wash/exe/mount.stream.js]: ; after return

File: lib/wash/exe/mount.stream.js (hunk 13 above)
Comment (ussuri) [inline lib/wash/exe/mount.stream.js]: I didn't understand about the `;`, but here's a couple others: 1) we use single quotes for strings; 2) I'd keep the vocabulary to a minimum by sticking to `'websocket'` instead of `'ws'`.

File: lib/wash/exe/mount.stream.js (hunk 14 above)
Comment (ussuri) [inline lib/wash/exe/mount.stream.js]: Same here

File: chrome_agent/js/chrome_agent.js (hunk 3 above)
Comment (ussuri) [inline chrome_agent/js/chrome_agent.js]: For completeness sake, the code here could stay in the ctor if `connectCompleter_` were added and the function you mention were to look like `connect() { return this.connectCompleter_.promise; }`.

File: lib/wash/exe/mount.stream.js (hunk 13 above)
Comment (rpaquay) [inline lib/wash/exe/mount.stream.js]: wrt to semicolon: `{return "ws"})` should be `{return 'ws';})`

Comment (ussuri) [inline lib/wash/exe/mount.stream.js]: Ah! :)

File: chrome_agent/js/chrome_agent.js (hunk 3 above)
Comment (umop) [inline chrome_agent/js/chrome_agent.js]: Agreed.  Or we could just keep the code in the constructor and add an onConnected Promise property to the class.  I think for simplicity sake, I'll go with the `connect()` method idea (except I'm calling it `listen()` because it's waiting for a connection), and return a promise with it.

Comment (umop) [inline chrome_agent/js/chrome_agent.js]: On second thought, I realized that `onConnected` is the only one that makes sense, since this is a listener which might be connected to multiple times (after being disconnected), so I switched out the promise for an event

File: chrome_agent/js/extension/background.js (hunk 4 above)
Comment (umop) [inline chrome_agent/js/extension/background.js]: Sounds good, except that this is doing the opposite (it is the agent, so it is listening for a connection from Axiom).  Also, I realized that because a listener can be connected to multiple times, this shouldn't be a promise, so I switched it to an event system (onConnected).

File: lib/axiom/fs/stream/extension_streams.js (hunk 6 above)
Comment (umop) [inline lib/axiom/fs/stream/extension_streams.js]: Well, ExtensionStream isa  PostMessageStream.  Someone could conceivably connect to a port directly if they wanted to, with an ExtensionStream instance.  Also, this would change the signature of the method (passing it an string / appId instead of a Port)

File: lib/axiom/fs/stream/extension_streams.js (hunk 7 above)
Comment (umop) [inline lib/axiom/fs/stream/extension_streams.js]: Same answer =)

File: lib/axiom/fs/stream/generic_streams.js (hunk 8 above)
Comment (umop) [inline lib/axiom/fs/stream/generic_streams.js]: Cool!  I decided to add a bit more checking (check to see if the class has been overridden before throwing the error.  That way we can call the `super` javascript equivalent without it throwing.

File: lib/axiom/fs/stream/generic_streams.js (hunk 9 above)
Comment (umop) [inline lib/axiom/fs/stream/generic_streams.js]: done

File: lib/axiom/fs/stream/post_message_streams.js (hunk 12 above)
Comment (umop) [inline lib/axiom/fs/stream/post_message_streams.js]: This was during experimenting with using a simple object for open so that I could extend upon it with other `open`s, before I realized that there should be separate open methods per subclass.

File: lib/axiom/fs/stream/generic_streams.js (hunk 11 above)
Comment (umop) [inline lib/axiom/fs/stream/generic_streams.js]: I agree.  GenericStreams was kind of a placeholder because I couldn't think of something better.  Done

File: lib/axiom/fs/stream/post_message_streams.js (hunk 10 above)
Comment (umop) [inline lib/axiom/fs/stream/post_message_streams.js]: done

File: lib/wash/exe/mount.stream.js (hunk 13 above)
Comment (umop) [inline lib/wash/exe/mount.stream.js]: (mostly) done.  The `'ws'` is inherited from the previous code (just refactored).  I'll leave that to @rpaquay.

File: lib/wash/exe/mount.stream.js (hunk 14 above)
Comment (umop) [inline lib/wash/exe/mount.stream.js]: `'ext'` is to match the `'ws'` from previous code.  Will leave that in @rpaquay's hands

**Comment by umop** [comment]
PTAL

File: gruntfile.js (hunk 5 above)
Comment (umop) [inline gruntfile.js]: I think that everything should be copied here.  Since this is an extension, we don't know what kinds of source might be here for sure (assets, etc).

**Comment by umop** [comment]
Cc @rginda

File: chrome_agent/background.html [hunk 15]
```diff
@@ -0,0 +1,12 @@
+<!doctype html>
```
Comment (rpaquay) [inline chrome_agent/background.html]: Is this file still needed now that `manifest.json` has been updated to contain a `background.scripts` entry?

File: lib/axiom/fs/stream/memory_stream_buffer.js [hunk 16]
```diff
@@ -83,4 +83,5 @@ MemoryStreamBuffer.prototype.close = function(error) {
   // This will fire an event chain:
   // this.writableStream.onClose -> this.onClose -> this.readableStream.onClose. 
   this.writableStream.close(error);
+  //this.onClose.fire();
```
Comment (rpaquay) [inline lib/axiom/fs/stream/memory_stream_buffer.js]: ?

**Comment by rpaquay** [comment]
lgtm after last 2 comments addressed.

File: chrome_agent/background.html (hunk 15 above)
Comment (umop) [inline chrome_agent/background.html]: Nope.  Thanks!  Done.

File: lib/axiom/fs/stream/memory_stream_buffer.js (hunk 16 above)
Comment (umop) [inline lib/axiom/fs/stream/memory_stream_buffer.js]: done
